rest-service/manager_rest/resource_manager.py

```diff
@@ -1176,16 +1177,39 @@ def prepare_deployment_nodes_for_storage(self,
     def _prepare_deployment_node_instances_for_storage(self,
                                                        deployment_id,
                                                        dsl_node_instances):
+
+        # The node_instance_index is the index of a node instance list for a
+        # node. It is used for serial operations on node instances of the
+        # same node. First we get the list of current node instances for the
+        # deployment.
+        deplyment_id_filter = self.create_filters_dict(
+            deployment_id=deployment_id)
+        dep_node_instances = self.sm.list(
+            models.NodeInstance,
+            filters=deplyment_id_filter,
+            get_all_results=True
+        )
+        # We build a dictionary in order to track the current index.
+        current_node_index = defaultdict(int)
+        for ni in dep_node_instances:
+            if ni.node_instance_index > current_node_index[ni.node_id]:
+                current_node_index[ni.node_id] = ni.node_instance_index
+
         node_instances = []
         for node_instance in dsl_node_instances:
             node = get_node(deployment_id, node_instance['node_id'])
+            # Update current node index.
+            node_instance_index = node_instance.get(
```

Comment:
I am not sure that the indexing going to be consistent

Reply:
@tyacbovi Added more tests also here to see how the index maintains consistency after deployment modification.